plugins/linux/iptablesplugin/descriptor/rulechain.go

```diff
@@ -0,0 +1,418 @@
+	ruleChainInterfaceDep = "interface-exists"
+
+	// minimum number of namespaces to be given to a single Go routine for processing
+	// in the Retrieve operation
+	minWorkForGoRoutine = 3
+)
+
+// A list of non-retriable errors:
+var (
+	// ErrCustomChainWithoutName is returned when the chain name is not provided for the custom iptables chain.
+	ErrCustomChainWithoutName = errors.New("iptables chain of type CUSTOM defined without chain name")
+
+	// ErrDefaultPolicyOnNonFilterRule is returned when a default policy is applied on a table different to FILTER.
+	ErrDefaultPolicyOnNonFilterRule = errors.New("iptables default policy can be only applied on FILTER tables")
+
+	// ErrDefaultPolicyOnCustomChain is returned when a default policy is applied on a custom chain, which is not allowed in iptables.
+	ErrDefaultPolicyOnCustomChain = errors.New("iptables default policy cannot be applied on custom chains")
+)
+
+// RuleChainDescriptor teaches KVScheduler how to configure Linux iptables rule chains.
+type RuleChainDescriptor struct {
+	log             logging.Logger
+	nsPlugin        nsplugin.API
+	scheduler       kvs.KVScheduler
+	ipTablesHandler linuxcalls.IPTablesAPI
+
+	// parallelization of the Retrieve operation
+	goRoutinesCnt int
+}
+
+// NewRuleChainDescriptor creates a new instance of the iptables RuleChain descriptor.
+func NewRuleChainDescriptor(
+	scheduler kvs.KVScheduler, ipTablesHandler linuxcalls.IPTablesAPI, nsPlugin nsplugin.API,
+	log logging.PluginLogger, goRoutinesCnt int) *RuleChainDescriptor {
+
+	return &RuleChainDescriptor{
+		scheduler:       scheduler,
+		ipTablesHandler: ipTablesHandler,
+		nsPlugin:        nsPlugin,
+		goRoutinesCnt:   goRoutinesCnt,
+		log:             log.NewLogger("ipt-rulechain-descriptor"),
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with the KVScheduler.
+func (d *RuleChainDescriptor) GetDescriptor() *adapter.RuleChainDescriptor {
+	return &adapter.RuleChainDescriptor{
+		Name:                 RuleChainDescriptorName,
+		NBKeyPrefix:          linux_iptables.ModelRuleChain.KeyPrefix(),
+		ValueTypeName:        linux_iptables.ModelRuleChain.ProtoName(),
+		KeySelector:          linux_iptables.ModelRuleChain.IsKeyValid,
+		KeyLabel:             linux_iptables.ModelRuleChain.StripKeyPrefix,
+		ValueComparator:      d.EquivalentRuleChains,
+		Validate:             d.Validate,
+		Create:               d.Create,
+		Delete:               d.Delete,
+		UpdateWithRecreate:   d.UpdateWithRecreate,
+		Retrieve:             d.Retrieve,
+		Dependencies:         d.Dependencies,
+		RetrieveDependencies: []string{ifdescriptor.InterfaceDescriptorName},
+	}
+}
+
+// EquivalentRuleChains is a comparison function for two RuleChain entries.
+func (d *RuleChainDescriptor) EquivalentRuleChains(key string, oldRCh, newRch *linux_iptables.RuleChain) bool {
+
+	// first, compare everything except the rules
+	oldCopy := proto.Clone(oldRCh).(*linux_iptables.RuleChain)
```

Comment:
I'm afraid that copying may have an impact on performance since the comparison function is called quite often.